modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/maven/AbstractMavenClassLoaderModelLoader.java

```diff
@@ -161,14 +166,42 @@ protected ClassLoaderModel createHeavyPackageClassLoaderModel(File artifactFile,
           .collect(toSet());
     }
 
-    loadUrls(artifactFile, classLoaderModelBuilder, bundleDependencies);
+    // TODO MULE-17114 retrieve this data from the json if present, if not then call this
+    populateNotImportingPackages(artifactFile, classLoaderModelBuilder, bundleDependencies);
+
     classLoaderModelBuilder.dependingOn(bundleDependencies);
 
     return classLoaderModelBuilder.build();
   }
 
+  protected void populateNotImportingPackages(File artifactFile, final ArtifactClassLoaderModelBuilder classLoaderModelBuilder,
+                                              Set<BundleDependency> bundleDependencies) {
+    final JarExplorer fileJarExplorer = buildJarExplorer();
+
+    // This is already filtering out mule-plugin dependencies,
+    // since for this case we explicitly need to consume the exported API from the plugin.
+    final List<URL> dependenciesArtifactsUrls = loadUrls(artifactFile, classLoaderModelBuilder, bundleDependencies);
+    for (URL dependencyArtifactUrl : dependenciesArtifactsUrls) {
+      final URI dependencyArtifactUri;
+      try {
+        dependencyArtifactUri = dependencyArtifactUrl.toURI();
+      } catch (URISyntaxException e) {
+        throw new MuleRuntimeException(e);
+      }
+
+      final JarInfo exploredJar = fileJarExplorer.explore(dependencyArtifactUri);
+
+      classLoaderModelBuilder.notImportingPackages(exploredJar.getPackages());
+      classLoaderModelBuilder.notImportingResources(exploredJar.getResources());
+    }
+  }
+
+  protected JarExplorer buildJarExplorer() {
```

Comment:
Looks like you want to mock using extension of this class. Cannot you include a JarExplorerFactory as a param instead?

Reply:
The wording of this comment reminded me of 
![image](https://user-images.githubusercontent.com/8010105/61005263-43a99d80-a33e-11e9-95cf-df3c17e5d994.png)